x/fungible/keeper/zrc20_methods.go

```diff
@@ -0,0 +1,301 @@
+func (k Keeper) ZRC20Allowance(
+	ctx sdk.Context,
+	zrc20ABI *abi.ABI,
+	zrc20Address, owner, spender common.Address,
+) (*big.Int, error) {
+	if zrc20ABI == nil {
+		return nil, fungibletypes.ErrZRC20NilABI
+	}
+
+	if crypto.IsEmptyAddress(owner) || crypto.IsEmptyAddress(spender) {
+		return nil, fungibletypes.ErrZeroAddress
+	}
+
+	if err := k.IsValidZRC20(ctx, zrc20Address); err != nil {
+		return nil, err
+	}
+
+	// function allowance(address owner, address spender)
+	args := []interface{}{owner, spender}
+	res, err := k.CallEVM(
+		ctx,
+		*zrc20ABI,
+		fungibletypes.ModuleAddressZEVM,
+		zrc20Address,
+		big.NewInt(0),
+		nil,
+		true,
+		true,
+		allowance,
+		args...,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if res.VmError != "" {
+		return nil, fmt.Errorf("EVM execution error calling allowance: %s", res.VmError)
+	}
+
+	ret, err := zrc20ABI.Methods[allowance].Outputs.Unpack(res.Ret)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ret) == 0 {
+		return nil, fmt.Errorf("no data returned from 'allowance' method")
+	}
+
+	allowanceValue, ok := ret[0].(*big.Int)
+	if !ok {
+		return nil, fmt.Errorf("ZRC20 allowance returned an unexpected type")
+	}
+
+	return allowanceValue, nil
+}
+
+// ZRC20BalanceOf checks the ZRC20 balance of a given EOA.
+func (k Keeper) ZRC20BalanceOf(
+	ctx sdk.Context,
+	zrc20ABI *abi.ABI,
+	zrc20Address, owner common.Address,
+) (*big.Int, error) {
+	if zrc20ABI == nil {
+		return nil, fungibletypes.ErrZRC20NilABI
+	}
+
+	if crypto.IsEmptyAddress(owner) {
+		return nil, fungibletypes.ErrZeroAddress
+	}
+
+	if err := k.IsValidZRC20(ctx, zrc20Address); err != nil {
+		return nil, err
+	}
+
+	res, err := k.CallEVM(
+		ctx,
+		*zrc20ABI,
+		fungibletypes.ModuleAddressZEVM,
+		zrc20Address,
+		big.NewInt(0),
+		nil,
+		true,
+		true,
+		balanceOf,
+		owner,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if res.VmError != "" {
+		return nil, fmt.Errorf("EVM execution error calling balanceOf: %s", res.VmError)
+	}
+
+	ret, err := zrc20ABI.Methods[balanceOf].Outputs.Unpack(res.Ret)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ret) == 0 {
+		return nil, fmt.Errorf("no data returned from 'balanceOf' method")
+	}
+
+	balance, ok := ret[0].(*big.Int)
+	if !ok {
+		return nil, fmt.Errorf("ZRC20 balanceOf returned an unexpected type")
+	}
+
+	return balance, nil
+}
+
+// ZRC20TotalSupply returns the total supply of a ZRC20 token.
+func (k Keeper) ZRC20TotalSupply(
+	ctx sdk.Context,
+	zrc20ABI *abi.ABI,
+	zrc20Address common.Address,
+) (*big.Int, error) {
+	if zrc20ABI == nil {
+		return nil, fungibletypes.ErrZRC20NilABI
+	}
+
+	if err := k.IsValidZRC20(ctx, zrc20Address); err != nil {
+		return nil, err
+	}
+
+	// function totalSupply() public view virtual override returns (uint256)
+	res, err := k.CallEVM(
+		ctx,
+		*zrc20ABI,
+		fungibletypes.ModuleAddressZEVM,
+		zrc20Address,
+		big.NewInt(0),
+		nil,
+		true,
+		true,
+		totalSupply,
+	)
+	if err != nil {
+		return nil, err
```

Comment:
consider wrapping  the error

Reply:
Wrapped all non-typed errors in 6adb5ec9bdbdf6b1239764fda7509a62bdf27ecb